### PR TITLE
feat(en): Fetch L1 batch root hashes from main node

### PR DESCRIPTION
**Pull request by slowli**
## What ❔

Implements a PoC for the treeless node mode, i.e., a mode in which a node doesn't maintain a Merkle tree locally, but rather downloads necessary data from a trusted peer (i.e., the main node). This extends to the case when a node is configured to start a Merkle tree after some time or immediately; the root hashes produced by the tree (which are expected to be slower than the tree data fetcher) will be used to double-check fetched hashes. Also, these hashes are checked by the consistency checker against L1 commitment transactions. 

## Why ❔

- Merkle tree recovery takes more than 50% of overall recovery process on the mainnet, so by introducing this mode, a recovering node can start operating a couple of hours earlier.
- Merkle tree introduces non-trivial operational requirements (~2.5 TB disk space / ~16 GB RAM on the mainnet), so by not running the tree, these requirements can be reduced.

## Checklist

- [x] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [x] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [x] Code has been formatted via `zk fmt` and `zk lint`.
- [x] Spellcheck has been run via `zk spellcheck`.

fix(state-keeper): Fix hanging up during shutdown if L1 batch root hashes are not being computed

### CODE REVIEW

File: core/lib/dal/src/blocks_dal.rs [hunk 1]
```diff
@@ -804,33 +804,12 @@ impl BlocksDal<'_, '_> {
         if update_result.rows_affected() == 0 {
             tracing::debug!("L1 batch #{number}: tree data wasn't updated as it's already present");
 
-            // Batch was already processed. Verify that existing hash matches
-            let matched: i64 = sqlx::query!(
```
Comment (slowli) [inline core/lib/dal/src/blocks_dal.rs:808]: Is there a particular reason the tree data was checked like this? I've simplified it just to use `get_l1_batch_tree_data()`.

File: core/lib/zksync_core/src/state_keeper/keeper.rs [hunk 2]
```diff
@@ -274,23 +274,50 @@ impl ZkSyncStateKeeper {
...
+                .await?
             {
-                return Ok(envs);
+                return Ok(params);
             }
         }
         Err(Error::Canceled)
     }
 
+    async fn wait_for_new_batch_env(
```
Comment (slowli) [inline core/lib/zksync_core/src/state_keeper/keeper.rs:293]: This fixes state keeper hanging up during node shutdown if the requested L1 batch root hash is not being computed. It was technically an issue even without introducing tree data fetcher; I've just encountered this problem during local testing and decided it makes sense to fix.

**Review by slowli — COMMENTED** [review]
To keep this PR reasonably small, I'd prefer to add integration testing in a separate PR.

File: core/bin/external_node/src/main.rs [hunk 3]
```diff
@@ -641,6 +642,16 @@ async fn init_tasks(
         None
     };
 
+    if components.contains(&Component::TreeFetcher) {
```
[the remark this answers is not in the document]
Comment (popzxc) [reply core/bin/external_node/src/main.rs:645]: Nah, I think it's fine.

File: core/lib/dal/src/blocks_dal.rs (hunk 1 above)
Comment (popzxc) [reply core/lib/dal/src/blocks_dal.rs:808]: Probably historical reasons(tm).

**Review by popzxc — DISMISSED** [review]

**Review by popzxc — APPROVED** [review]

**Pull request merged**